### DISCUSSION TITLE
Compare with `typy`

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#111] [General]
We should check out what [@cyrus- has done with `typy`](https://twitter.com/neurocy/status/1134502111964815361), to see what we can learn from it. From a first glance, it seems like they are overloading getitem to do typing, which maybe we can avoid now with annotations.

But it would be good to get a better sense of the core of their system and see how it compares.

**Comment by cyrus-** [comment]
Hi!

We parse the AST so no run time operator overloading. But yes syntactically we were using the getitem notation. This was before there were annotations on assignments so you may be able to avoid that.

**Comment by saulshanabrook** [comment]
Ah, cool, yeah we have talked about parsing the AST or the bytecode, but have been avoiding it so far.

I have been looking into [this work](http://www.cs.purdue.edu/homes/rompf/papers/decker-preprint201811.pdf) a bit as well, which transforms the AST to turn control flow into function calls:

<img width="533" alt="Screen Shot 2019-06-03 at 12 32 56 PM" src="https://user-images.githubusercontent.com/1186124/58818327-bfa41d80-85fb-11e9-9f23-d408ff407ecf.png">


Seems like eventually we will have to do something like the turning a control flow graph into a [value state dependence graph](https://www.microsoft.com/en-us/research/publication/optimizing-compilation-value-state-dependence-graph/), so we can do replacements on the functional representation.

Or just not let people use Python's control flow and force them to use your own functional API. It's what Tensorflow did for a while!
